core/utils/codel.h

```diff
@@ -0,0 +1,301 @@
+  article http://queue.acm.org/detail.cfm?id=2209336
+
+  It provides an active queue management to help prevent bufferbloat by dropping
+  queue entries at an increasing rate if the delay in the queue is above the
+  target queue delay. The equation used to calculate drop intervals is based on TCP
+  throughput response to drop probability.
+*/
+struct Codel {
+  double delay_target;  // the delay that codel will adjust for
+  double window;        // minimum time before changing state
+  // the time at which codel will change state to above target(0 if below)
+  double time_above_target;
+  double time_till_drop;  // the next time codel will drop
+  // the number of packets dropped while delay has been above target
+  uint32_t drop_count;
+  uint8_t dropping;  // whether in dropping state(above target for window)
+  llring* ring;      // queue
+  void (*drop_func)(void*);  // the function to call to drop a value
+};
+
+/*
+  Initalizes the codel queue. Takes a pointer to malloc'd memory in the heap to
+  initialize the codel into, function to call on dropped objects, the number of
+  slots to have in the queue, the target queue delay_target, the minimum amount
+  of time before codel will change from not dropping to dropping. Returns 0 on
+  success and negative error code on failure
+*/
+static inline int CodelInit(Codel* c, void (*drop_func)(void*),
+                            uint32_t slots = DEFAULT_SLOTS,
+                            double target = DEFAULT_TARGET,
+                            double window = DEFAULT_WINDOW) {
+  int ret;
+  int bytes = llring_bytes_with_slots(slots);
+  c->ring = static_cast<llring*>(mem_alloc_ex(bytes, alignof(llring), 0));
+  if (!c->ring) {
+    return -ENOMEM;
+  }
+
+  ret = llring_init(c->ring, slots, 1, 1);
+  if (ret) {
+    mem_free(c->ring);
+    return -EINVAL;
+  }
+
+  c->delay_target = target;
+  c->window = window;
+  c->dropping = 0;
+  c->drop_count = 0;
+  c->time_above_target = 0;
+  c->time_till_drop = get_epoch_time() + window;
+  c->drop_func = drop_func;
+
+  return 0;
+}
+
+/*
+  A wrapper class around the pointer to track the delay time and whether
+  the packet can be dropped after dequeueing.
+*/
+struct Wrapper {
+  void* ptr;         // a pointer to be enqueued
+  double timestamp;  // a timestamp of the time the pointer was enqueued
+  uint8_t drop;  // used during dequeue to determin if the packet can be dropped
+};
+
+/*
+  This algorithm gets a fast approximate inverse sqrt
+  it is taken from https://cs.uwaterloo.ca/~m32rober/rsqrt.pdf
+
+  Takes the value to get the inverse square root for and
+  returns the inverse square root of d.
+*/
+static inline double InvSqrt(double d) {
+  double d2 = d * 0.5;
+  uint64_t magic;
+  memcpy(&magic, &d, sizeof(magic));
+  magic = 0x5fe6eb50c7b537a9 - (magic >> 1);
+  double d3;
+  memcpy(&d3, &magic, sizeof(magic));
+  d3 = d3 * (1.5 - d2 * d3 * d3);
+  return d3;
```

Comment:
Please do not use this method -- it has been obsolete since at least the early 2000s. Just use pow(x, -2.0) or one of the gcc builtins. This method is not a good idea.